### PR TITLE
[shardformer] Pipeline/whisper

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This PR adds whisper pipeline forward, policies and tests, including debugging

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/whisper.py [hunk 1]
```diff
@@ -223,13 +229,175 @@ def add_lm_head_policy(self, base_policy):
...
+
+        # the number of stages distributed between encoder and decoder is optmized in this way:
+        # num_encoder_stages = argmin(abs(num_encoder_layers / encoder_stages - num_decoder_layers / decoder_stages))
+        #                   s.t. num_encoder_stages + num_decoder_stages = num_stages, num_encoder_stages >= 1, num_decoder_stages >= 1
+        def objective(num_encoder_stages):
+            return abs(num_encoder_layers / num_encoder_stages - num_decoder_layers / (num_stages - num_encoder_stages))
+
+        num_encoder_stages = 0
+        optimal_diff = 2**31 - 1
+        for i in range(1, num_stages):
```
Comment (ver217) [inline colossalai/shardformer/policies/whisper.py:261]: You could use `np.argmin()` directly

File: colossalai/shardformer/policies/whisper.py [hunk 2]
```diff
@@ -223,13 +229,175 @@ def add_lm_head_policy(self, base_policy):
...
+
+        if stage_manager.stage < decoder_starting_stage:
+            # current stage is in whisper's encoder
+            if stage_manager.is_first_stage():
+                held_layers.append(encoder.embed_positions)
+                held_layers.append(encoder.conv1)
+                held_layers.append(encoder.conv2)
+            if stage_manager.stage == decoder_starting_stage - 1:
+                held_layers.append(encoder.layer_norm)
+            held_layers.extend(encoder.layers[start_idx:end_idx])
```
Comment (ver217) [inline colossalai/shardformer/policies/whisper.py:324]: Is it possible that a stage have both encoder layer and decoder layer?

Comment (ver217) [reply colossalai/shardformer/policies/whisper.py:324]: Can we handle them together?

Comment (CjhHa1) [reply colossalai/shardformer/policies/whisper.py:320]: in that case the forward will be complicated as we need to transfer encoder outputs, I prefer to delay this case?

File: colossalai/shardformer/policies/whisper.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/whisper.py]: done

File: colossalai/shardformer/policies/t5.py [hunk 3]
```diff
@@ -228,13 +229,9 @@ def distribute_t5_layers(num_encoder_layers: int, num_decoder_layers: int,
...
-        num_encoder_stages = 0
-        optimal_diff = 2**31 - 1
-        for i in range(1, num_stages):
-            attempt = objective(i)
-            if attempt < optimal_diff:
-                num_encoder_stages = i
-                optimal_diff = attempt
+        diff = range(1, num_stages)
+        num_encoder_stages = np.argmin([objective(i) for i in diff]) + 1
+        num_decoder_stages = num_stages - num_encoder_stages
```
Comment (ver217) [inline colossalai/shardformer/policies/t5.py:234]: It seems that this line is duplicated.

File: colossalai/shardformer/policies/t5.py [hunk 4]
```diff
@@ -228,13 +229,9 @@ def distribute_t5_layers(num_encoder_layers: int, num_decoder_layers: int,
...
 
-        num_encoder_stages = 0
-        optimal_diff = 2**31 - 1
-        for i in range(1, num_stages):
-            attempt = objective(i)
-            if attempt < optimal_diff:
-                num_encoder_stages = i
-                optimal_diff = attempt
+        diff = range(1, num_stages)
+        num_encoder_stages = np.argmin([objective(i) for i in diff]) + 1
```
Comment (ver217) [inline colossalai/shardformer/policies/t5.py:233]: `diff` is not meaningful enough. You can just move it in for-loop.

File: colossalai/shardformer/policies/whisper.py (hunk 2 above)
Comment (ver217) [reply colossalai/shardformer/policies/whisper.py:320]: OK. Please add a TODO here

Comment (CjhHa1) [reply colossalai/shardformer/policies/whisper.py:320]: done

File: colossalai/shardformer/policies/t5.py (hunk 3 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/t5.py]: removed

File: colossalai/shardformer/policies/t5.py (hunk 4 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/t5.py]: ok! , I put this in the for loop

**Review by ver217 — APPROVED** [review]

**Pull request merged**